[master] Core/Scripts: Baleroc Encounter implementation

Krudor opened the pull request:
**Changes proposed**:
- Implement Baleroc Encounter

**Target branch(es)**: 6x

**Issues addressed**: Closes  #11558

**Tests performed**: Build & vigorous in-game testing

**Known issues**:
- (Trash) Magmakin doesn't explode when reaching the player (Not addressed because of suspected core issue with two similiar distance calculation functions returning different results)
- (Trash) Flame Archon's cancel their own spellcasts if they are simply being kited (Core issue)
- When Baleroc is dual wielding his normal weapons, he swings with both weapons simultaneously. (Creature dual wielding not implemented properly?)
- Inferno Strike and Decimating Strike are not cast on the tank. (Requires implementation of SPELL_AURA_OVERRIDE_AUTOATTACK_WITH_MELEE_SPELL)
- Players do not gain Vital Flame when healing targets with Indendiary Soul. (Unknown cause, proc triggering Vital Flame is no longer functioning after porting from a cataclysm rev (UNK), assuming it is core related)
- Vital Flame will, even if applied not increase healing done (Requires implementation of SPELL_AURA_359?)
- Shards of Torment has on one occasion been seen to cast Torment on dead players (Possibly just when the player had been killed, problems in the core?)
- Shards of Torment turn to face the player it is casting Torment on (Creature does not have NPC_FLAG2_DISABLE_TURN or any other flag that could intervene with this, can only blame it on the core)
- Shards of Torment's Torment channel tether snaps at the channel spells cast range of 15 yards, when it should break at 30 yards (trigger spell range, fix core?)
- No loot tables.

Creature immunities are not verified.

Let's give it another shot.

Krudor committed:
Core/Instances: Add functionality for modifying instance entrance locations

Instances have since MoP (Possibly even Cataclysm) been able to set new
locations for players when they zone into the instance.

A theoretic example of this could be after a certain boss has been
defeated, a new entrance location is set to bring players entering the
instance closer to the relevant content.
A real example would be Siege of Orgrimmar changing entrance locations
many times as the raid progresses.

Added SetEntranceLocation for setting entrance locations that will be
saved to DB the next time the instance is saved, and
SetTemporaryEntranceLocation for when you don't want the value to be
saved to the database.

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into 6.x

Krudor committed:
Revert "Core/Instances: Add functionality for modifying instance entrance locations"

This reverts commit c95ef343b121572cc9bdc4dab5ea57907786ee5b.

Krudor committed:
First Baleroc update revision

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into baleroc_revisited

# Conflicts:
#	src/server/scripts/Spells/spell_generic.cpp

Krudor committed:
Fix to an interesting previous merge conflict commit

Krudor committed:
Baleroc updates

Krudor committed:
Baleroc Torment 6.x update & removed unused enums

Krudor committed:
Line up spacing

DDuarte commented on src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:
```diff
@@ -0,0 +1,1250 @@
+            }
+            
+            WorldObject* _target1;
+            WorldObject* _target2;
```
If possible store guids instead of pointers

DDuarte commented on src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:
```diff
@@ -0,0 +1,1250 @@
+                return target;
+            }
+
+            static WorldObject* GetRandomContainerElement(std::list<WorldObject*> list)
```
Pass by const ref

DDuarte commented on src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:
```diff
@@ -0,0 +1,1250 @@
+                if (!list.empty())
+                    return Trinity::Containers::SelectRandomContainerElement(list);
+
+                return static_cast<WorldObject*>(nullptr);
```
static_cast not required

Krudor committed:
Code style changes

Krudor committed:
Fix Travis CI errors

Krudor committed:
Travis CI Round 2

tkrokli commented on src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:
```diff
@@ -0,0 +1,1249 @@
+                uint32 auraCnt = GetCaster()->GetAuraCount(SPELL_VITAL_SPARK);
+                int32 healingPct = sSpellMgr->GetSpellInfo(SPELL_VITAL_SPARK)->GetEffect(EFFECT_0)->BasePoints * auraCnt;
+
+                if (uint32(GetAura()->GetEffect(EFFECT_0)->GetAmount()) < healingPct)
```
```
/home/travis/build/TrinityCore/TrinityCore/src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:1180:73: fatal error: 
      comparison of integers of different signs: 'uint32' (aka 'unsigned int')
      and 'int32' (aka 'int') [-Wsign-compare]
  ...if (uint32(GetAura()->GetEffect(EFFECT_0)->GetAmount()) < healingPct)
         ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ ^ ~~~~~~~~~~
```

MitchesD commented:
nice stuff, hope it will be merged possibly even with few issues, rather than let it freeze for another 2 years

Naios commented on src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:
```diff
@@ -0,0 +1,1249 @@
+
+            void OnApply(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
+            {
+                ENSURE_AI(boss_baleroc::boss_balerocAI, GetTarget()->GetAI())->EquipWeapon(EQUIP_INFERNO_BLADE);
```
Use `DoAction` instead of AI casts please

Naios commented on src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:
```diff
@@ -0,0 +1,1249 @@
+                    return;
+
+                targets.sort(Trinity::ObjectDistanceOrderPred(GetCaster(), true));
+                targets.resize(1);
```
Use a random container select instead of a full sort please.

Krudor committed:
Use DoAction over ENSURE_AI & Travis CI Round 3

Aokromes commented on src/server/scripts/Kalimdor/Firelands/firelands.cpp:
```diff
@@ -0,0 +1,111 @@
+{
+    new spell_firelands_smouldering();
+    new spell_firelands_smouldering_aura();
+}
```
missing newline.

tkrokli replied:
Yup, needs an extra LF/CR in local editor.

Krudor replied:
I'll fix this and make some other adjustments after I'm done wrapping up one of my other internal commits :)

Temppus commented on src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:
```diff
@@ -0,0 +1,1274 @@
+                    ObjectGuid player;
+                    uint32 tormented;
+                }
+                _sharedThePain[25];
```
Consider using something like:
std::set<ObjectGuid, uint32> shareThePainSet;

No need to define own struct, flexible length, no need to iterate -> find on set.

Shauren replied:
always default to std::unordered_map unless for some reason you need it sorted by key

Shauren commented on src/server/game/Spells/SpellMgr.cpp:
```diff
@@ -3407,6 +3407,19 @@ void SpellMgr::LoadSpellInfoCorrections()
+            // FIRELANDS SPELLS
+            //
+            case 99253: // Torment Searcher
+                spellInfo->RangeEntry = sSpellRangeStore.LookupEntry(EFFECT_RADIUS_15_YARDS);
```
This doesn't work at all. You are mixing range (distance between caster and spell destination, defined per spell) with radius (distance between spell destination and target, defined per effect)

Krudor replied:
Thanks for pointing this out, I'll take care of this in my next update.

Krudor committed:
More updates

Fixed torment searcher spell range code (Thanks Shauren).
Updated old Shared the Pain achievement tracker container.
Fixed an issue where if Shards of Torment never aquired a target to
begin with, they would never start pulsating AOE damage.
Fixed an issue where Shards of Torment did not despawn if Baleroc
entered evade mode.

Krudor committed:
Removed changes unrelated to this commit by request

Krudor commented:
Not really, the map will return a std::pair, not the key value, I could use the auto keyword though.

Krudor committed:
Simplified code and fixed a compilation error

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/6.x' into baleroc_revisited

Krudor committed:
Updated an iterator to const ref

Krudor commented:
Moved berserk spell fix to new PR #17561

Temppus commented on src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:
```diff
@@ -0,0 +1,1257 @@
+enum Misc
+{
+    EQUIP_DEFAULT                   = 1,
+    EQUIP_INFERNO_BLADE             = 2,
```
Does 71138 + 71082 entries cause Baleroc's weapon to shine when he empowers his blades ?
If not, use these entries:

```
enum ItemEntries
{
    NORMAL_BLADE_ENTRY        = 71055,
    INFERNO_BLADE_ENTRY       = 94155,
    DECIMATION_BLADE_ENTRY    = 94157,
};
```

Krudor replied:
They already work as they should, what item ids are those that you provided?

tkrokli replied:
Ignore the suggestion, none of those 3 item IDs exist in the DB anyway and only 1 was found on wowhead: http://www.wowhead.com/item=94155/staff-of-antonidas

Temppus replied:
Sorry, maybe old entries from 4.3.4.

Aokromes commented:
What's missing here?

Ulduar commented:
About 359 aura -  i see 
``` CPP
ID - 99263 Целительное пламя
=================================================
Description: В вас проснулся могущественный целительный огонь, значительно повышающий эффективность ваших исцеляющих заклинаний, применяемых к целям, находящимся под воздействием эффекта "Пламя славы" на $d.

ToolTip: Эффективность исцеляющих заклинаний, применяемых к целям, находящимся под воздействием эффекта "Пламя славы", повышена на $w1%.
Category = 0, SpellIconID = 2327, activeIconID = 0, SpellVisual = (0,0)
Family SPELLFAMILY_GENERIC, flag [0] 0x00000000 [1] 0x00000000 [2] 0x00000000 [3] 0x00000000

SpellSchoolMask = 4 (SPELL_SCHOOL_MASK_FIRE)
DamageClass = 0 (SPELL_DAMAGE_CLASS_NONE)
PreventionType = 0 (SPELL_PREVENTION_TYPE_NONE)
=================================================
Attributes: 0x80000000 (SPELL_ATTR0_CANT_CANCEL)
AttributesEx8: 0x00001000 (SPELL_ATTR8_AURA_SEND_AMOUNT)
=================================================
Spell Level = 0, base 0, max 0

Category = 0
DispelType = 0 (DISPEL_NONE)
Mechanic = 0 (MECHANIC_NONE)
SpellRange: (Id 1) "Self Only":
    MinRangeNegative = 0, MinRangePositive = 0
    MaxRangeNegative = 0, MaxRangePositive = 0

Cast time (Id 1) = 0,00
Duration: ID (8)  15000, 0, 15000
Interrupt Flags: 0x00000000, AuraIF 0x00000000, ChannelIF 0x00000000
Chance = 0, charges - 0
=================================================
Effect 0: Id 6 (SPELL_EFFECT_APPLY_AURA) DIFFICULTY_NONE
BasePoints = 10
Targets (1, 0) (TARGET_UNIT_CASTER, NO_TARGET)
Aura Id 359 (SPELL_AURA_359), value = 10, misc = 22 (22), miscB = 0, periodic = 0
```
10 - this is amount, understand, misc = 22 - AuraState?
Maybe 359 aura - this is SPELL_AURA_MOD_HEALING_PERCENT_DONE_VERSUS_AURASTATE?
What do you think?

Ulduar commented:
Suggest (not testing)

``` CPP
float Unit::SpellHealingPctDone(Unit* /*victim*/, SpellInfo const* spellProto) const
{
    // For totems pct done mods are calculated when its calculation is run on the player in SpellHealingBonusDone.
    if (GetTypeId() == TYPEID_UNIT && IsTotem())
        return 1.0f;

    // No bonus healing for potion spells
    if (spellProto->SpellFamilyName == SPELLFAMILY_POTION)
        return 1.0f;

    float DoneTotalMod = 1.0f;

    // bonus against aurastate
    AuraEffectList const& mHealingDonePctAurastate = GetAuraEffectsByType(SPELL_AURA_MOD_HEALING_PERCENT_DONE_VERSUS_AURASTATE);
    for (AuraEffectList::const_iterator i = mHealingDonePctAurastate.begin(); i != mHealingDonePctAurastate.end(); ++i)
        if (victim->HasAuraState(AuraStateType((*i)->GetMiscValue())))
            AddPct(DoneTotalMod, (*i)->GetAmount());
    
    // Healing done percent
    AuraEffectList const& mHealingDonePct = GetAuraEffectsByType(SPELL_AURA_MOD_HEALING_DONE_PERCENT);
    for (AuraEffectList::const_iterator i = mHealingDonePct.begin(); i != mHealingDonePct.end(); ++i)
        AddPct(DoneTotalMod, (*i)->GetAmount());

    return DoneTotalMod;
}
```
``` CPP
&AuraEffect::HandleNoImmediateEffect,                         //359 SPELL_AURA_MOD_HEALING_PERCENT_DONE_VERSUS_AURASTATE implemented in Unit::SpellHealingPctDone
```

Ulduar commented:
@Krudor what about my suggestion?

Krudor commented:
I should have gotten back to your earlier, I don't know what aurastates are so I can't review your suggestion.

Krudor commented:
hi.

tkrokli commented:
@Krudor : in your opinion, is this PR ready for merging?

Krudor commented:
I really don't see why it shouldn't be.
I have nothing new to commit to this PR, so either merge or close in my opinion.

Shauren commented:
What about that large TODO list you wrote?

tkrokli commented:
It would be a good idea to edit the list and use tick boxes instead of bullet points:
- [ ] This to-do is not complete yet.
- [X] This to-do is complete.
```
- [ ] This to-do is not complete yet.
- [X] This to-do is complete.
```

Krudor commented:
The large list of TODO's are not something that I will implement in this PR.
Alot of them are issues/unimplemented features in the core that are ultimately unrelated to this PR.
Fixes to these issues and features would fix this encounter without even touching the script.

Many times when I make scripts for certain encounters I always reach a point where there are missing features that I have to sidestep and fix/try to implement, and those features in turn might require something else that needs coding as well.
It's very easy to lose focus on what you're doing, you can't fix everything.
Baleroc is a tricky case, it's a small fight with multiple technological obstacles to overcome.

As for loot, it certainly is related to this, but is ultimately something I'm not interested in doing.
I spent a couple of days calculating accurate drop rates for Firelands a long time ago, but eventually lost the data at a point, and the thought of doing it again honestly makes me sick.
If anyone wants to pitch in with the loot tables, that'd be great, but after that's been done I believe this is ready for merge.

If you require all features used encounters to be implemented, then that is sad for me to hear and I'll probably stop contributing if that is the case. I've scripted ALOT of encounters, and I've come to realize that it is nearly impossible to finish scripting an encounter if you don't draw the line for what an encounter implementation consists of. There's always going to be core features, unrelated to the script that break the encounter that you simply have to ignore for the moment.

Krudor commented on src/server/scripts/Kalimdor/Firelands/instance_firelands.cpp:
```diff
@@ -40,8 +46,59 @@ class instance_firelands : public InstanceMapScript
+                }
+                else if (type == DATA_BALEROC && state == IN_PROGRESS)
+                    if (GameObject* door = instance->GetGameObject(BalerocDoorGUID))
+                        door->SetGoState(GO_STATE_READY);
```
I don't think so, the behavior of the door is a little bit too special for that to handle.

joschiwald replied:
how should it work? it is not exactly clear to me, different states in OnGameObjectCreate and SetBossState

tkrokli commented:
OK, thanks for clarifying.

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/master' into baleroc_revisited

Krudor committed:
Legion update

Added some apparently missing header files.
Minor code style changes.
Fixed a shards of torment spell casting issue.
Fixed a shards of torment spell speech logic issue.

Krudor commented:
I will add loot tables and check on balerocs immunity mask shortly, after thats done there should be no reason for why this shouldn't be merged.

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/master' into baleroc_revisited

Krudor committed:
Added loot and adjusted mechanic immune mask

Shared loot table doesn't appear to work as intended.
Whenever the shared loot item table is chosen, no loot item is
distributed.
Drop rates should be pretty good approximates.

Krudor commented:
Mechanic mask has been adjusted.

Loot looks good except for the shared loot table not working as intended.
Whenever an item is calculated to be from the Firelands shared loot table, no item is distributed.
If someone's good with loot templates, feel free to take a look at the sql.

Baleroc's loot template references a reference_loot_template that includes Baleroc's loot items. One of the loot entries in that template is a reference to another reference_loot_template containing the shared boss loot, it should have a 5% chance to proc that reference and give us a random item from that template.

My PR for a Ragnaros script #18058 is currently experiencing the same issue.

Keader commented on src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:
```diff
@@ -0,0 +1,1257 @@
+                scheduler.Update(diff);
+
+                if (!me->HasUnitState(UNIT_STATE_CASTING))
+                    DoMeleeAttackIfReady();
```
DoMelleAttackIfReady() , already check if is casting

Krudor replied:
Oh yeah, thanks

Keader commented on src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:
```diff
@@ -0,0 +1,1257 @@
+                {
+                    DoCastAOE(SPELL_MAGMA_CONDUIT);
+                    if (me->GetMap()->Is25ManRaid())
+                        DoCastAOE(SPELL_MAGMA_CONDUIT);
```
it works without trigger cast?

Krudor replied:
Instant cast, shouldn't collide unless they have cast times.

Keader commented on src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:
```diff
@@ -0,0 +1,1257 @@
+                scheduler.Schedule(Seconds(5), [this](TaskContext context)
+                {
+                    DoCastAOE(SPELL_MAGMA_CONDUIT);
+                    if (me->GetMap()->Is25ManRaid())
```
if (Is25ManRaid())

Keader commented on src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:
```diff
@@ -0,0 +1,1257 @@
+                        {
+                            Talk(SAY_ENRAGE);
+                            Talk(EMOTE_ENRAGE);
+                            DoCast(me, SPELL_BERSERK);
```
DoCastSelf(SPELL_BERSERK);

Krudor replied:
Nice, haven't seen or used that one.

Keader commented on src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:
```diff
@@ -0,0 +1,1257 @@
+
+            void HandleScript(SpellEffIndex /*effIndex*/)
+            {
+                if (GetHitUnit())
```
GetHitUnit() only can be null in this hook, if spell hit objects

Keader commented on src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:
```diff
@@ -0,0 +1,1257 @@
+        {
+            PrepareSpellScript(spell_baleroc_decimating_strike_SpellScript);
+
+            bool Validate(SpellInfo const* /*spellInfo*/)
```
missing override

Keader commented on src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:
```diff
@@ -0,0 +1,1257 @@
+                SetHitDamage(healthPctDmg < flatDmg ? flatDmg : healthPctDmg);
+            }
+
+            void Register()
```
missing override

Keader commented on src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:
```diff
@@ -0,0 +1,1257 @@
+            }
+        };
+
+        SpellScript* GetSpellScript() const
```
missing override

Keader commented on src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:
```diff
@@ -0,0 +1,1257 @@
+
+            void OnRemove(AuraEffect const* /*aurEff*/, AuraEffectHandleModes /*mode*/)
+            {
+                GetTarget()->ToPlayer()->RemoveAurasDueToSpell(SPELL_COUNTDOWN_VISUAL_LINK);
```
GetTarget()->RemoveAurasDueToSpell(SPELL_COUNTDOWN_VISUAL_LINK);

Krudor committed:
Review adjustments and spell validation fixes

tkrokli commented on src/server/scripts/Kalimdor/Firelands/boss_baleroc.cpp:
```diff
@@ -0,0 +1,1333 @@
+/*
+ * Copyright (C) 2008-2016 TrinityCore <http://www.trinitycore.org/>
```
` * Copyright (C) 2008-2017 `

Krudor replied:
Hm, was pretty sure I changed this, perhaps it was on the Ragnaros script :)

tkrokli commented on src/server/scripts/Kalimdor/Firelands/firelands.cpp:
```diff
@@ -0,0 +1,111 @@
+*
+* You should have received a copy of the GNU General Public License along
+* with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
```
Alignment - use the same 1st column spacing as in the header of `boss_baleroc.cpp` + change 2016 to 2017 in line 2.

Krudor committed:
Changed copyright header year to 2017

Krudor committed:
Merge remote-tracking branch 'refs/remotes/TrinityCore/master' into baleroc_revisited

Aokromes commented:
Conflicting files
src/server/game/Spells/SpellMgr.cpp
src/server/scripts/Kalimdor/Firelands/firelands.cpp
src/server/scripts/Kalimdor/Firelands/firelands.h

tkrokli commented:
I would have updated the conflicting files if this PR was tagged as allowed for TC members to edit.
@Krudor, do you want to solve the conflicts yourself, or can the TC members get PR access?
